### DISCUSSION TITLE
Save on `eglot-rename` in otherwise-unmodified files?

### DISCUSSION THREAD

**Discussion by jbr** [joaotavora/eglot#1383] [Q&A] [answered]
Is there a way to have Eglot save buffers when it performs an edit to them? I'd like to be able to `M-x eglot-rename RET [new name] RET` and have my code compile regardless of how many files contained that symbol. Currently I switch to each buffer that was modified and manually save them, which is very unergonomic.

Alternatively, could Eglot offer a `eglot-rename-and-save` that prompts to save any file that previously had modifications before the Eglot-rename and saves otherwise-unmodified files without interaction?

**Comment by joaotavora** [comment]
There is no such functionality.  I doubt it should be provided by Eglot though.  Have you tried `save-some-buffers`?  Maybe there should be a `project-save-some-buffers`: that's what you should probably run before you `M-x project-compile`  or `project-save-some-buffers`.  in fact doesn't `M-x compile` already do that for you?

**Reply by joaotavora** [reply, marked as answer]
Of course, maybe you don't compile your project with `M-x compile`.  That's fine, but adding a `save-some-buffers` into your loop, perhaps as a "before" advice, might make sense.

CCing @dgutov might have ideas here.  Personally I'm stepping away from Emacs deve as much as possible btw.

**Reply by jbr** [reply]
Yeah, I don't compile in emacs. I was hoping there was a hook/config snippet to automatically save-some-buffers within the eglot workspace (not all open buffers necessarily) after eglot-rename. I'll see if I can write something for myself like that and post it here after a bit.

Thanks for eglot, @joaotavora!

**Reply by dgutov** [reply]
It seems to me like `eglot-rename` should start with a forced saving of related buffers - because the server process likely only has access to file contents on disk (or operates on them anyway). Someone please correct me if I'm wrong here.

Regarding the "save after" - I'm also of the same opinion as the OP, that the rename operation should end with a new state, also saved to disk. But that's a UI-level concern which we could/should make configurable, and would be a part of the alternative renaming/refactoring UI we discussed on the emacs-devel mailing list.

If the buffers are all saved before the rename, it should be fine to save them after.

**Reply by joaotavora** [reply]
> because the server process likely only has access to file contents on disk (or operates on them anyway). Someone please correct me if I'm wrong here.

You're wrong :-)

Server has a view of the buffer, doesn't need to be saved.  That's why "didChange" events are exchanged.

**Reply by joaotavora** [reply]
>  and would be a part of the alternative renaming/refactoring UI we discussed on the emacs-devel mailing list.

Yes.  

I'd probably configure it myself to "not save".  Or at least, "don't save the thing that's in front of me", if that was somehow an option.  

Tangent, IIRC Eglot's behavior is to visit files targeted by an LSP edit if they are not visited.  I think I'd also like the option to have it change those files on disk, i.e. don't visit into buffers if there isn't one already.  Lots of options here.

**Reply by dgutov** [reply]
> Server has a view of the buffer, doesn't need to be saved. That's why "didChange" events are exchanged.

I'm aware of those, but do the language servers operate on them consistently in "rename" operations, disregarding the contents on disk? I.e. all/most of the LSP implementations.

If so, the current Eglot behavior is very logical, but a "let's keep things saved on disk" user option would probably make things easier for many users anyway.

**Reply by dgutov** [reply]
> Tangent, IIRC Eglot's behavior is to visit files targeted by an LSP edit if they are not visited. I think I'd also like the option to have it change those files on disk, i.e. don't visit into buffers if there isn't one already. Lots of options here.

Right. I would also add that to the list of requirements to the new "rename things" interface.

**Reply by joaotavora** [reply]
> I'm aware of those, but do the language servers operate on them consistently in "rename" operations, disregarding the contents on disk?

Yes, at least the ones I use daily.

> I.e. all/most of the LSP implementations.

There are hundreds.  But the LSP idea as I've always understood it is to have a virtual file system.  I.e don't require saves and don't even _require_ an actual file system (although some servers use the disk of course for indexing other parts).

I.e. you could be using the server through the network where it never sees your local files, you just inform it of the things that you're doing (changing files, deleting them, creating them) and it has a view of the state of the whole "workspace" without having physical access to a large part of it. Some servers allow that, though I've rarely seen it used. 

Personally, I like Eglot's behaviour. It's how an editor should work. Saving the file without really needing to throws away a perfectly usable `revert-buffer`.  And saving the file may trigger other things and could even be slow.  Anyway I always prefer saves to be on my explicit command.

**Comment by dgutov** [comment]
Speaking of `project-save-some-buffers`, there is no such command at the moment, but Emacs 28.1 added function `save-some-buffers-root` that can be used as the value of `save-some-buffers-default-predicate`.

That alters the behavior of the command globally.

EDIT: But to use this kind of functionality temporarily in a program, you could bind the variable in a `let` form.

**Comment by larrasket** [comment]
This is one of the features  that I really miss after migrating from LSP. It does not only affect `eglot-rename` but any server action that involves other buffers. Even compiling in Emacs does not solve the issue: my compile command for example might just be running some `.sh` that seems irrelevant to Emacs' project structure, so nothing in that root is being saved after the compile command, and I always have to save-some-buffers.
